### ISSUE TITLE
TCVOM - Blurry Edges, Am I doing something wrong? 

### ISSUE THREAD

**Issue by bfan1256** [yunkezhang/TCVOM#10]
Hey there, trying to run this on my own individual dataset, but I can't seem to reproduce the sharpe edges found in your results in the paper (even for single frame of 1280x720)

Using `EvalModel` `VMN-FBA`

## Alpha Matte Result
![Image 11-23-22 at 9 47 PM](https://user-images.githubusercontent.com/15917894/203691957-1cb18044-60ae-44c4-985e-ee82b319c4d9.jpg)

## Input Frame
![Screen Shot 2022-11-25 at 11 32 48 AM](https://user-images.githubusercontent.com/15917894/204033758-7db8dc7b-2ee6-4785-92f8-1dbb47487996.png)

## Trimap
![Screen Shot 2022-11-25 at 11 30 28 AM](https://user-images.githubusercontent.com/15917894/204033494-6250667e-957f-4a69-9bb6-a704332c5052.png)

## Alpha Matte
![Screen Shot 2022-11-25 at 11 29 09 AM](https://user-images.githubusercontent.com/15917894/204033502-2f1e2d8b-cc5a-4ad8-9697-5206ad99c239.png)

**Comment by yunkezhang** [comment]
It seems like your alpha matte does not correspond to the input image. Can you provide the corresponding alpha matte or input image? 

Judging from the input frame though, the target object has a blueish glow that gradually blends into the background. To me, it's normal that the alpha matte has a smooth edge. 
720p is not a very high resolution by today's standard (as in our dataset all videos are in 1080p). The quality of the image (especially the sharpness around the edge) is not very high either. This could also contribute to smoothness.

**Comment by bfan1256** [comment]
Attaching the alpha matte associated here: 
![Screen Shot 2022-11-25 at 11 29 09 AM](https://user-images.githubusercontent.com/15917894/204033357-185dfeb8-3f3f-4b4b-bc42-f3c24f49617c.png)

**Note: Frame with the blue is alpha matte displayed layered on top of the frame**

Yes, I suspect that the low resolution could probably cause this? But I also do see an element where the trimap sizing causes very different results. When we set a very narrow trimap, we get a good result, but a medium to wide trimap, and it has a faint glow as output from TCVOM.

The question persists though since it does seem like the matte is a bit beyond the output of the bunny itself. Is this expected behavior? Oftentimes, it seems that TCVOM produces a map that is almost perfectly aligned/smoothed with the edge especially in instances with decent contrast even with a generous trimap like the peacock demo video

**Comment by yunkezhang** [comment]
Thanks for the explanation and the image!
Yes, the trimap size is crucial to the matting performance. As seen in our paper, "narrow" and "wide" trimaps have different metrics even with the same object in the same frame.

As for the "glow", I still think it might be related to the image type and quality. Our model hasn't been trained on any 3D-animated videos, and the bunny is relatively small in the frame (which is different from our training dataset where the foreground object is very salient).

TBH, the target object in this frame has little detail and very high background and foreground contrast. In this case, using traditional non-DL approaches like sampling-based or affinity-based algorithms with proper cropping might be more desirable.

**Comment by bfan1256** [comment]
Sounds good! I'll mark this issue as closed! Thanks for the recommendations. Just curious to hear your thoughts. A lot of work being done in video object segmentation (VOS) - seems like masks and trimaps are super similar, wonder if there's some interesting corollary work, considering too that VOS systems can almost all run in RT, where it seems like TCOVM is around ~2-3s / frame

**Comment by yunkezhang** [comment]
In recent years there are a lot of video matting methods that could achieve real-time performance and they do not require trimap as input (e.g. MODNet and BackgroundMattingV2), albeit with some limitations such as requiring a specific type of video (e.g. the foreground object must be human).

TCVOM on the other hand is more focused on temporal consistency. The network actually forwards 3 frames during inference, utilizing the previous and the next frame's features to predict a more accurate alpha matte for the current frame (it can definitely be optimized through caching features, but it's out of the scope for this demo). Thus, the inference speed is not very fast, especially with a relatively large base method like FBA.
